### PR TITLE
EVG-12676 remove only-api config

### PR DESCRIPTION
**Pull request by john-m-liu**
Sorry for the huge pr, but I didn't want this to get partially deployed since we would be without a way to update service users
- Remove the only-api config section, replaced by a single boolean flag for whether the system supports service users
- Change the only-api auth manager to create an instance of itself if this flag is set
- Change the source of service users to be the users collection, rather than the list in the old config
- Create rest routes to read/update/delete service users. These were code generated and since they're the first ones, the architecture is subject to change (feedback is welcome)
- Create CLI commands to call these routes
- To deploy, change "only_api" to "allow_service_users" in the multi auth config before deploying this commit

### CODE REVIEW

**Comment by john-m-liu** [comment]
Oh and there are still a couple bugs, but I would say this is at least 95% finalized and ready to look at

File: auth/only_api_test.go [hunk 1]
```diff
@@ -1,404 +0,0 @@
-package auth
```
Comment (Kimchelly) [inline auth/only_api_test.go:1]: Can we still test this?

File: config_auth.go [hunk 2]
```diff
@@ -200,10 +171,8 @@ func (c *AuthConfig) ValidateAndDefault() error {
 				catcher.NewWhen(c.Github == nil, "GitHub settings cannot be empty if using in multi auth")
 			case AuthNaiveKey:
 				catcher.NewWhen(c.Naive == nil, "Naive settings cannot be empty if using in multi auth")
-			case AuthOnlyAPIKey:
-				catcher.NewWhen(c.OnlyAPI == nil, "OnlyAPI settings cannot be empty if using in multi auth")
 			default:
-				catcher.Errorf("unrecognized auth mechanism '%s'", kind)
+				grip.Errorf("unrecognized auth mechanism '%s'", kind)
```
Comment (Kimchelly) [inline config_auth.go:175]: Why was this changed?

File: operations/admin.go [hunk 3]
```diff
@@ -645,3 +646,114 @@ func adminDistroExecute() cli.Command {
...
+			cli.StringFlag{
+				Name:  displayNameFlag,
+				Usage: "the display name of the service user",
+			},
+			cli.StringSliceFlag{
+				Name:  rolesFlag,
+				Usage: "The roles for the user. Must be the entire list of roles to set",
+			},
+		},
+		Before: mergeBeforeFuncs(
```
Comment (Kimchelly) [inline operations/admin.go:704]: Get rid of unnecessary `mergeBeforeFuncs`.

File: operations/admin.go [hunk 4]
```diff
@@ -645,3 +646,114 @@ func adminDistroExecute() cli.Command {
 		},
 	}
 }
+
+func getServiceUsers() cli.Command {
+	return cli.Command{
+		Name:  "get-service-users",
+		Usage: "prints all service users",
+		Before: mergeBeforeFuncs(
```
Comment (Kimchelly) [inline operations/admin.go:654]: Get rid of unnecessary `mergeBeforeFuncs`.

File: operations/admin.go [hunk 5]
```diff
@@ -44,6 +44,7 @@ func Admin() cli.Command {
 			toMdbForLocal(),
 			updateRoleCmd(),
 			adminDistroExecute(),
+			updateServiceUser(),
```
Comment (Kimchelly) [inline operations/admin.go:47]: Include `getServiceUsers` and `deleteServiceUser`.

File: rest/route/user.go [hunk 6]
```diff
@@ -438,3 +438,105 @@ func (h *userRolesPostHandler) Run(ctx context.Context) gimlet.Responder {
...
+		return errors.New("'id' must be specified")
+	}
+
+	return nil
+}
+
+func (h *serviceUserDeleteHandler) Run(ctx context.Context) gimlet.Responder {
+	err := h.sc.DeleteServiceUser(h.username)
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(err)
```
Comment (Kimchelly) [inline rest/route/user.go:509]: Same question about internal error responder.

File: rest/route/user.go [hunk 7]
```diff
@@ -438,3 +438,105 @@ func (h *userRolesPostHandler) Run(ctx context.Context) gimlet.Responder {
...
+	return nil
+}
+
+func (h *serviceUserPostHandler) Run(ctx context.Context) gimlet.Responder {
+	if h.u == nil {
+		return gimlet.NewJSONInternalErrorResponse("error reading request body")
+	}
+	err := h.sc.AddOrUpdateServiceUser(*h.u)
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(err)
```
Comment (Kimchelly) [inline rest/route/user.go:475]: Shouldn't this be InternalErrorResponder? Otherwise it'll return bad request.

File: rest/route/user.go [hunk 8]
```diff
@@ -438,3 +438,105 @@ func (h *userRolesPostHandler) Run(ctx context.Context) gimlet.Responder {
 
 	return gimlet.NewJSONResponse(struct{}{})
 }
+
+type serviceUserPostHandler struct {
```
Comment (Kimchelly) [inline rest/route/user.go:442]: Can we have tests for these?

File: auth/auth.go [hunk 9]
```diff
@@ -45,8 +45,8 @@ func LoadUserManager(settings *evergreen.Settings) (gimlet.UserManager, evergree
 	if authConfig.Github != nil {
 		return makeGithubManager(settings, authConfig.Github)
 	}
-	if authConfig.OnlyAPI != nil {
-		return makeOnlyAPIManager(authConfig.OnlyAPI)
+	if authConfig.AllowServiceUsers {
```
[the remark this answers is not in the document]
Comment (Kimchelly) [reply auth/auth.go:48]: Still same question about this.

[the remark this answers is not in the document]
Comment (john-m-liu) [reply auth/auth.go:48]: I think I did this. The only API section is back, but doesn't do anything

File: auth/only_api.go [hunk 10]
```diff
@@ -85,35 +25,38 @@ func NewOnlyAPIUserManager(config *evergreen.OnlyAPIAuthConfig) (gimlet.UserMana
...
+		return nil, errors.Wrap(err, "unable to create user cache")
 	}
 	return cached.NewUserManager(cache)
 }
 
 // findOnlyAPIUser finds an API-only user by ID and verifies that it is a valid
 // user against the list of authoritative valid users.
-func findOnlyAPIUser(id string, validIDs []string, validUsers []gimlet.User) (*user.DBUser, error) {
+func findOnlyAPIUser(id string) (*user.DBUser, error) {
+	validUsers, err := user.FindServiceUsers()
```
[the remark this answers is not in the document]
Comment (john-m-liu) [reply auth/only_api.go:52]: done

File: auth/only_api.go [hunk 11]
```diff
@@ -10,68 +9,24 @@ import (
...
-			SystemRoles: u.Roles,
-			OnlyAPI:     true,
-		}
-
-		validUsers = append(validUsers, dbUser)
-		validIDs = append(validIDs, u.Username)
+func NewOnlyAPIUserManager() (gimlet.UserManager, error) {
+	validUsers := []gimlet.User{}
+	validIDs := []string{}
+	users, err := user.FindServiceUsers()
```
[the remark this answers is not in the document]
Comment (Kimchelly) [reply auth/only_api.go]: I actually meant that it's not necessary to call `FindServiceUsers` at all (L52-62 and L73-77) because you can just assume that a bot will have `only_api: true` in the DB.

File: auth/only_api_test.go (hunk 1 above)
Comment (Kimchelly) [reply auth/only_api_test.go:1]: Gonna bump this. It seems correct for now, but I don't want someone to change something later and it breaks all our service users.

Comment (john-m-liu) [reply auth/only_api_test.go]: Oops, restored this and removed the tests that are updating data

File: auth/only_api.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (john-m-liu) [reply auth/only_api.go]: I told myself I would never do this, but alas I committed the change but never pushed. Should be fixed now

**Comment by john-m-liu** [comment]
Looks like the index already exists, so just adding it to indexes.js for posterity
